README
Geo::Coordinates::OSGB
Geo::Coordinates::OSTN02
----------------------

Routines to convert latitude and longitude coordinates to and from transverse
Mercator grid references, that are optimized for use with the British
national grid.

The OSGB module provides routines to convert between latitude/longitude
coordinates based on the British Ordnance Survey's standard geoid model and
the British National Grid.  The module includes some useful extra routines to
parse and format grid references in a variety of popular forms (including
which Landranger map your point appears on).

The second module provides an implementation of the standard transformation
tables that are now used by the British Ordnance Survey to define the
National Grid from GPS surveys.  This module provides accurate conversion
between the British National Grid and latitude/longitude coordinates based on
the global WGS84 geoid model (the one used by all GPS devices, and popular
mapping tools such as Google Earth).  The OSTN02 data is rather large (6M
bytes), and will take a few microseconds to load even on a new machine,
but once loaded conversions should be reasonably quick.

INSTALLATION

This package installs in the regular Perl way.  You don't need the C compiler
and as far as I know there are no platform dependencies.  Either use your
local package tool, or just try

  [sudo] perl -MCPAN -e "install Geo::Coordinates::OSGB" 
  
For the die hards, the package can also be installed by
hand; just unzip it to some temporary home, then do the usual thing:

        Perl Makefile.PL
        make
        make test
        [sudo] make install

The last tests called by "make test" *will* generate some harmless warnings.
The full suite of tests takes less than 2 seconds on my machine. 
You may or may not need sudo depending on how you have set up your system.

DEPENDENCIES

None.

COPYRIGHT AND LICENCE

Copyright (C) 2002-2013 Toby Thurston

OSTN02 transformation data is freely available but remains Crown Copyright (C) 2002

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

All feedback gratefully received.


Changes
Revision history for Perl extension Geo::Coordinates::OSGB

2.06  October 2013

    - removed precise URL for theory paper - the OS keep moving it about, so just Google for the name
    - removed non-ascii characters from POD in OSGB.pm to avoid encoding issue
    

2.05  31 August 2013

    - Put blank lines back in OSTN02 to prevent pod2html taking an age during install
    - Refactored OSGB and OSTN02 lightly to conform more closely to "Perl Best Practices"
    - corrected corner coordinate data for Landranger sheets 1 and 185

2.04  6 November 2008

    - Rework search of OSTN02 data to improve performance and reliability
    - Add range checking of OSTN02 coordinates
    - Compress OSTN02 data to hexadecimal form to speed up loading
    - Change selection of data in 03random.t test to avoid off-the-sheet errors
    - Minor tidy up changes in OSGB
    - Complete overhaul of the documentation
    - removed old "coco.pl" example script
    - added new example script to show all "000 000" grid references

2.03 Not released

2.02 22 September 2008

    - Removed irritating degree sign from verbose test message (no functional change)
    - Make parse_landranger_grid behave better in scalar context
    - Improve useability of parse_grid
    - Correct GR of SW corner of Landranger Sheet 3
    - Further optimize handling of OSTN02 data to prevent data parsing at pod2man time
    - Improve documentation
    - More example scripts (with notes)

2.01 6 September 2007

    - Optimized handling of OSTN02 data to reduce memory requirements *and* speed it up.

2.00 4 September 2007

    - New version with routines to do adjustment to and from latitude and
      longitude coordinates based on the (now ubiquitous) WGS84 model, for
      more compatible use with GPS receivers and Google Earth and similar
      tools.

    - Includes not only fast approximate adjustment as used in most
      coordinate tools and devices, but also the definitive-but-slow OSTN02
      transformation based on OSGB data

    - Step version number

    - POD completely revised


1.07 7 September 2005

       - corrected link in POD

       - various minor documentation corrections

1.06 16 May 2005

       - general clean up

       - made the format and parse routines external to avoid spaghetti connections

       - added support for grid references in GPS format (eg TQ 32342 24312)

1.05 29 March 2005

       - Not released

1.04  29 September 2004

        - corrected a bug in grid2ll, where long/lat values with 0 degrees
          were always printed as positive in scalar form The problem was that
          -1 * 0 = +0 in Perl (as it says in the Camel book somewhere).

        - added simple coco.pl script to provide command line coordinate
          conversions.

        - actually implemented the long-promised map2ll and ll2map functions
          as well as map2grid and grid2map for completeness.  With these
          functions you can convert to and from coordinates in a more useable
          form, such as "OS Sheet 43; 875123".

        - added a demonstration script --- plotmaps.pl --- which will produce
          a pretty index of the Landranger maps, showing the GB coast line
          and the large grid squares.  The output is in PostScript.  Use
          GhostScript to process it.  A PDF version of the output is also
          included.

        - added a useful file with all the names and other details of the
          Landranger Sheets

1.03  18 December 2002
        - changed name for CPAN release
        - improved ll2grid function
        - new ll2map and map2ll functions


1.02  November 13, 2002
        - No changes to code, just tidied the documentation

0.01  Fri Mar  2 19:05:50 2001
        - original version; created by h2xs 1.19


ll_to_bng.pl
#! /usr/bin/perl -w

# Toby Thurston ---  6 May 2009
# Parse LL and show as National Grid ref

use strict;
use warnings;

use Geo::Coordinates::OSGB qw/
        parse_ISO_ll
        ll_to_grid
        shift_ll_from_WGS84
        format_ll_trad
        format_grid_trad
        format_grid_landranger/;

if ( @ARGV == 0 ) {
    die "Usage: $0 lat lon\n"
}

my ($lat, $lon);
if ( @ARGV == 1 ) {
    ($lat, $lon) = parse_ISO_ll($ARGV[0]);
} else {
    ($lat, $lon) = @ARGV;
}

my ($gla, $glo, undef) = shift_ll_from_WGS84($lat, $lon);
my ($ge, $gn) = ll_to_grid($gla, $glo);
my ($e,  $n)  = ll_to_grid($lat, $lon);

print "Your input: @ARGV\n";
printf "is %s\n", scalar format_ll_trad($lat, $lon);

printf "== %d %d from OSGB  (%s)\n", $e, $n, scalar format_grid_landranger($e, $n);
printf "or %d %d from WGS84 (%s)\n", $ge, $gn, scalar format_grid_landranger($ge, $gn);


make_kml_index.pl
#! /usr/bin/perl -wc

use strict;
use warnings;
use Readonly;

# Create a KML file to show all the LR sheets in Google Earth

warn "**************************************************************************************\n";
warn "***********  Experimental output!  May not work in your version of Google Earth ******\n";
warn "***********  Experimental output!  May not work in your version of Google Earth ******\n";
warn "***********  Experimental output!  May not work in your version of Google Earth ******\n";
warn "***********  Experimental output!  May not work in your version of Google Earth ******\n";
warn "**************************************************************************************\n";

# approach:
# using OSGB::CC get all the SW corner locations for each sheet in m
# workout the coordinates of each corner (by adding 40km)
# and translate to ll WGS84

# write KML header
# for each square write a polygon into the KML
# write KML trailer

use Geo::Coordinates::OSGB qw(grid_to_ll parse_landranger_grid format_grid_GPS);
use Geo::Coordinates::OSTN02 qw(OSGB36_to_ETRS89);
use Getopt::Std;

sub grid2ll {
    my $e = shift;
    my $n = shift;
    my ($x,$y,undef) = OSGB36_to_ETRS89($e,$n,0);
    my ($lat, $lon) = grid_to_ll($x,$y,'ETRS89');
    return ($lat, $lon);
}

my @sheets = ();
Readonly my $sheet_size => 40_000; # in metres
Readonly my $sheet_count => 204; # in metres

for my $s (1..$sheet_count) {
    my ($swe, $swn) = parse_landranger_grid($s);
    push @sheets, { number => $s,
                    sw => [ grid2ll( $swe             , $swn             ) ],
                    nw => [ grid2ll( $swe             , $swn+$sheet_size ) ],
                    ne => [ grid2ll( $swe+$sheet_size , $swn+$sheet_size ) ],
                    se => [ grid2ll( $swe+$sheet_size , $swn             ) ],
                  };
}


print <<'END_HEADER';
<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://earth.google.com/kml/2.0">
<Document>
<name>OSGB_Landranger_index.kml</name>
<Style id="normalLRMaps">
<PolyStyle id="MagentaOutline">
      <color>bb00bbff</color>
      <fill>0</fill>
      <outline>1</outline>
</PolyStyle>
</Style>
<Style id="selectLRMaps">
<PolyStyle id="MagentaOutline">
      <color>bb00bbff</color>
      <fill>0</fill>
      <outline>3</outline>
</PolyStyle>
</Style>
<StyleMap id="LRMaps">
  <Pair>
    <key>normal</key>
    <styleUrl>#normalLRMaps</styleUrl>
  </Pair>
  <Pair>
    <key>highlight</key>
    <styleUrl>#selectLRMaps</styleUrl>
  </Pair>
</StyleMap>
END_HEADER

for my $s (@sheets) {
    print "<Placemark>\n";
    print "<name>Sheet $s->{number}</name>\n";
    print "<styleURL>#LRMaps</styleURL>\n";
    print "<description>Sheet $s->{number}</description>\n";
    print "<Polygon><outerBoundaryIs><LinearRing><coordinates>\n";
    print join(',', reverse(@{$s->{sw}}), 0) , "\n";
    print join(',', reverse(@{$s->{nw}}), 0) , "\n";
    print join(',', reverse(@{$s->{ne}}), 0) , "\n";
    print join(',', reverse(@{$s->{se}}), 0) , "\n";
    print join(',', reverse(@{$s->{sw}}), 0) , "\n";
    print "</coordinates></LinearRing></outerBoundaryIs></Polygon>\n";
    print "</Placemark>\n";
}


print <<'END_FOOTER';
</Document>
</kml>
END_FOOTER


plotmaps.pl
#! perl -w
use strict;
#
# A program to produce a post-script plot of all the Landranger sheets,
# complete with the GB coast line and the 100km grid sqaure letters.
# Toby Thurston ---  7 Oct 2008
#
use Geo::Coordinates::OSGB "ll_to_grid", "parse_landranger_grid", "format_grid_GPS";
use Getopt::Std;
our $opt_a = 4;
getopt('a'); # paper size...
$opt_a = 4 unless $opt_a =~ /\A\d\Z/ && $opt_a < 6;


my @sheets = ();
my %squares = ();

my $extreme_south = 200000;
my $extreme_north = 0;
my $extreme_west  = 200000;
my $extreme_east  = 0;

my $sheet_size = 40_000;

my $GR_pattern = qr/^([A-Z][A-Z])(\d\d\d)(\d\d\d)$/;

for my $sheet (1 .. 204) {

    # get the grid coordinates for the SW corner of the map
    my ($x, $y) = parse_landranger_grid($sheet);

    # put this sheet into the list of sheets
    push @sheets, sprintf "%f %f moveto currentpoint %d dup rectstroke %d dup rmoveto (%s) cshow\n",
                          $x/1000, $y/1000, $sheet_size/1000, $sheet_size/2000, $sheet;


    # which grid squares does this sheet touch?
    for my $xx ($x, $x+$sheet_size-100) {
        for my $yy ($y, $y+$sheet_size-100) {
            my ($sq, $e, $n) = format_grid_GPS($xx,$yy);
            next if defined $squares{$sq};
            $squares{$sq} = sprintf "%f %f moveto currentpoint 100 100 rectstroke 50 40 rmoveto (%s) cshow\n",
                                      ($xx-$e)/1000, ($yy-$n)/1000, $sq;

        }
    }

    # keep track of bounding box
    $extreme_east = $x if $x > $extreme_east;
    $extreme_west = $x if $x < $extreme_west;
    $extreme_north = $y if $y > $extreme_north;
    $extreme_south = $y if $y < $extreme_south;

}

# allow for most north-eastern sheets
$extreme_east  += $sheet_size;
$extreme_south += $sheet_size;
$extreme_north += $sheet_size*4; # allow for the top most grid square ...

my $ew_range = $extreme_east-$extreme_west;
my $ns_range = $extreme_north-$extreme_south;

# size of the paper in mm
my ($width, $height) = (148,210);

$opt_a == 4 && (($width, $height) = (210,297));
$opt_a == 3 && (($width, $height) = (297,420));
$opt_a == 2 && (($width, $height) = (420,594));
$opt_a == 1 && (($width, $height) = (594,840));
$opt_a == 0 && (($width, $height) = (840,1188));

# convert to points and deduct margins (20pts)
$width  = int($width  / 25.4 * 72 + 0.5) - 40;
$height = int($height / 25.4 * 72 + 0.5) - 40;

my $urx = $width  + 20;
my $ury = $height + 20;

my $scale = 1000 * $width / $ew_range;
my $h_scale = 1000 * $height / $ns_range;
$scale = $h_scale if $h_scale < $scale;

print << "PREAMBLE";
%!PS-Adobe-3.0 EPSF-3.0
%%Creator: $0
%%Copyright: (C) 2008 Toby Thurston
%%Title:(Index to the Landranger Sheets)
%%CreationDate: (17 Jan 2008)
%%BoundingBox: 20 20 $urx $ury
%%Pages: 1
%%EndComments
%%BeginSetup
/Large { /Helvetica 24 selectfont } def
/small { /Helvetica 8 selectfont } def
/cshow { /s exch def s stringwidth pop neg 2 div dup 0 exch 0 rmoveto s show rmoveto } def
%%EndSetup
%%Page: 1 1
%%BeginPageSetup
/pgsave save def
%%EndPageSetup
small
$scale dup scale
170 30 translate
-120 0 850 1280 rectclip

PREAMBLE

# a file of longitude & latitude data in "matlab" format
open GB, "<gb-coastline.dat";
my $cmd = "moveto";
print "gsave .6 .6 1 setrgbcolor\n";
while (<GB>) {
    chomp;
    my ($lon, $lat) = split;
    if ( $lon eq '#' ) {
        print "stroke\n";
        $cmd = 'moveto'
    }
    else {
        my ($e, $n) = ll_to_grid($lat,$lon);
        printf "%.3f %.3f %s\n", $e/1000, $n/1000, $cmd;
        $cmd = 'lineto' if ($cmd eq 'moveto');
    }
}
close GB;
print "grestore\n";

print "gsave .6 1 .6 setrgbcolor\n";
for my $lon (-9 .. 2) {
    my ($e, $n) = ll_to_grid(49.95, $lon);
    printf "%.3f %.3f moveto gsave 2 0 rmoveto small ($lon) show /degree glyphshow grestore\n", $e/1000, $n/1000;
    for my $lat (500 .. 609) {
        ($e,$n) = ll_to_grid($lat/10, $lon);
        printf "%.3f %.3f lineto\n", $e/1000, $n/1000;
    }
    print "gsave 2 0 rmoveto small ($lon) show /degree glyphshow grestore stroke\n";
}

for my $lat (51 .. 60) {
    my ($e, $n) = ll_to_grid($lat, -9.2);
    printf "%.3f %.3f moveto gsave -2 2 rmoveto small ($lat) show /degree glyphshow grestore\n", $e/1000, $n/1000;
    for my $lon (-91 .. 22) {
        ($e,$n) = ll_to_grid($lat, $lon/10);
        printf "%.3f %.3f lineto\n", $e/1000, $n/1000;
    }
    print "gsave 0 2 rmoveto small ($lat) show /degree glyphshow grestore stroke\n";
}
print "grestore\n";

print "gsave Large 1 .6 .6 setrgbcolor\n";
print $squares{$_} for sort keys %squares;
print "grestore\n";

print ".3 setlinewidth\n";
print @sheets;
print "pgsave restore showpage\n%%EOF\n";
